Chatkit/Auth/TokenProvider.swift

```diff
@@ -0,0 +1,137 @@
+                retryAttempts: 3,
+                retryDelayMs: 1000
+            )
+        )
+    }
+
+    internal static func testTokenHTTPCallProvider(instanceLocator: String, userId: String) -> TokenProvider {
+        // TODO: Implement:
+        // extract host and instanceId from instanceLocator
+        let host = "us1.pusherplatform.io"
+        let instanceId = "UNIMPLEMENTED"
+        let path = "/services/chatkit_token_provider/v1/\(instanceId)/token"
+
+        return HTTPCallTokenProvider(
+            method: "POST",
+            host: host,
+            path: path,
+            headers: nil,
+            queryParams: [ "user_id": userId ]
+        )
+    }
+}
+
+/// CachingTokenProvider delegates the fetching of tokens to an underlying `delegateProvider`, retrying on failure
+public class CachingTokenProvider: TokenProvider {
+    private let delegate: TokenProvider
+    private var token: String?
+
+    public init(delegateProvider: TokenProvider) {
+        self.delegate = delegateProvider
+    }
+
+    public func fetchToken(completionHandler: @escaping (TokenProviderResult) -> Void) {
+        // TODO: Implement:
+        // if token is nil or expired, delegate
+        // if call to delegate is in progress, attach to its completion handler
+        // return token
+
+        self.delegate.fetchToken(completionHandler: completionHandler)
+    }
+}
+
+/// RetryingTokenProvider delegates the fetching of tokens to an underlying `delegateProvider`, maintains a reference to the last fetched token and ensures
+/// that the delegate is only called when necessary:
+public class RetryingTokenProvider: TokenProvider {
+    public init(delegateProvider: TokenProvider, retryAttempts: UInt, retryDelayMs: UInt) {}
+
+    public func fetchToken(completionHandler: @escaping (TokenProviderResult) -> Void) {
+        // TODO: Implement:
+        // fetch token from delegate, with retries on failure
+    }
+}
+
+/// HTTPCallTokenProvider makes calls to a specified HTTP endpoint and expects to receive a token from it.
+///
+/// This class will make an HTTP request for every call to `fetchToken`, and in almost all cases it should be wrapped with caching and retry logic. See `httpTokenProvider(...)` for a factory method which returns a wrapped instance of this class.
+///
+/// - Parameters:
+///     - method: The HTTP method to use in the call, e.g. `GET`, `POST`
+///     - host: The host name to be called, e.g. `example.com`
+///     - path: The path component of the URL to be called, e.g. `/tokens`
+///     - headers: An optional map of headers to include in the request. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+///     - queryParams: An optional map of query parameters to include in the request URL. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+public class HTTPCallTokenProvider: TokenProvider {
```

Comment:
Why do we want to expose this type? Having it as an implementation detail behind the factory method wouldn't be better? I thought that's the motivation for the factory method.

Reply:
Class cluster perhaps?
https://developer.apple.com/library/archive/documentation/General/Conceptual/CocoaEncyclopedia/ClassClusters/ClassClusters.html

Reply:
For advanced usage (as yet undocumented), we want people to be able to make different implementations or vary different levels of the stack (network, retries, caching). The factory methods are just conveniences for the common use case, which is why these types are accessible

Reply:
Still they could copy internal implementation (it's open source) and use it as a starting point for implementing their custom TokenProvider. Designing for extension is hard and establishes very intimate API contract ("is" as opposed to "uses"). Indeed already designing a good well established API just for instantiating those generic implementations is not trivial. I think we should keep the API surface minimal, only for the bits we're confident about and happy to commit to maintain as part of the API contract.